keps/sig-network/4444-service-traffic-distribution/README.md

```diff
@@ -1114,3 +1225,51 @@ distribution when used in conjunction.
   pod distribution across zones, maximizing the likelihood that the
   `trafficDistribution` can be satisfied and reduce (although not completely
   eliminate) chances of overload for a single zone.
+
+#### Making Traffic Distribution directly aware of Topology Spread Constraints
+
+Rather than using Topology Spread Constraints with `PreferClose`, we
+could have traffic distribution make use of Topology Spread
+Constraints directly (either by changing the semantics of
+`PreferClose` in this case, or by adding a new distribution mode such
+as `FollowTopologyConstraints`).
+
+In this case, when the EndpointSlice controller decided that a service
+should used TSC-based traffic distribution, it would add a new
+`ForNodeSelector` hint to the EndpointSlices.
+
+For example, if a pod had:
+
+```
+  topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        foo: bar
+```
+
+and the pod was on a node with the label
+`topology.kubernetes.io/zone: zone-a`, then the Endpoint corresponding
+to that Pod would be given:
+
+```
+  hints:
+    forNodeSelector:
+      matchLabels:
+        topology.kubernetes.io/zone: zone-a
+```
+
+kube-proxy would then use this hint to know that it should only use
+that Endpoint if it was on a node that matched the given label
+selector. (The use of a full label selector rather than a single value
+is to handle the case of endpoints with multiple topology
+constraints.)
```

Comment:
Overall, this idea seems like it's worth exploring. I like how it leverages existing features of Kubernetes. 

A use case that I think needs to be figured out, is when multiple constraints are set.
We often use:
```
  topologySpreadConstraints: &topologySpreadConstraints
    - maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
      matchLabelKeys:
        - pod-template-hash
      nodeTaintsPolicy: Honor
    - maxSkew: 1
      topologyKey: kubernetes.io/hostname
      whenUnsatisfiable: DoNotSchedule
      matchLabelKeys:
        - pod-template-hash
      nodeTaintsPolicy: Honor
```

I assume in this example, if the destination Pod was on the same node as the origin, then `FollowTopologyConstraints` would use that Pod, rather than spreading it across the whole zone.

Reply:
Yes, as the text is currently written, clients on the same node as an endpoint would pick a same-node endpoint, and clients on nodes with no endpoint would pick a same-zone endpoint. Maybe that's the wrong heuristic. I'm not really that familiar with how TopologySpreadConstraints is used...